### PR TITLE
RFC to change dependency loading process

### PR DESCRIPTION
**Pull request by aleksei-fedotov**
### Description 
Add RFC about loading dependencies by module name only.

### Type of change

- [ ] bug fix - _change that fixes an issue_
- [ ] new feature - _change that adds functionality_
- [ ] tests - _change in tests_
- [ ] infrastructure - _change in infrastructure and CI_
- [X] documentation - _documentation update_

### Tests

- [ ] added - _required for new features and some bug fixes_
- [X] not needed

### Documentation

- [ ] updated in # - _add PR number_
- [ ] needs to be updated
- [X] not needed

### Breaks backward compatibility
- [ ] Yes
- [X] No
- [ ] Unknown

### CODE REVIEW

**Review by pavelkumbrasev — DISMISSED** [review]
The proposal loos good to me!
@vossmjp please take a look

**Comment by aleksei-fedotov** [comment]
@egfefey please take a look as well.

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+1. Introduce signature verification step to the run-time dependency loading
+   process.
+2. Introduce the ~TBB_VERIFY_DEPENDENCY_SIGNATURE~ compilation option that would
+   enable signature verification, and set it ~ON~ by default.
+3. Update documentation to include information about new
+   ~TBB_VERIFY_DEPENDENCY_SIGNATURE~ flag.
+4. Pass ~LOAD_LIBRARY_SAFE_CURRENT_DIRS~ flag to the ~LoadLibraryEx~ calls so
+   that current working directory is excluded from the list of directories in
+   which the system loader looks when trying to find and resolve dependency.
+5. Change dependency loading approach to load by module names only.
```
Comment (akukanov) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:89]: Which parts of this proposal apply to Windows only, Linux only, and both?

Comment (aleksei-fedotov) [reply rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org]: Clarified where parts belong.

**Review by akukanov — APPROVED** [review]

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 2]
```diff
@@ -0,0 +1,104 @@
+#+title: Loading Dependencies By Module Name
+
+* Introduction
+There is a well-known attack that implies loading of a malicious dependency
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:4]:
```suggestion
There is a well-known attack that involves loading of a malicious dependency
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+instead of the original one without notice to the party that does this loading.
+In the industry it is usually called /DLL injection/ or /DLL preloading attack/
+and it is mostly associated with the Windows platform as it is known to be
+particularly vulnerable to this kind of attack.[1] One of the recommendations
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:8]:
```suggestion
particularly vulnerable to this kind of attack [1]. One of the recommendations
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+that safeguards against this type of attack is to specify fully qualified path
+to a dependency.[2]
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:10]:
```suggestion
to a dependency [2].
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+Historically, oneTBB loads its optional dependencies during its initialization
+process when these dependencies can be used for the first time. The way oneTBB
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:13]:
```suggestion
process when these dependencies are used for the first time. The way oneTBB
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+does this is by building full paths to their dependencies using the path where
+oneTBB library resides by itself. It is the only sensible path which can be
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:15]:
```suggestion
the oneTBB library itself resides. It is the only sensible path which can be
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+obtained by oneTBB, whose usage conditions are not known at the time of
+development. The purpose is to minimize the risk of a DLL injection attack issue
+so that only certain paths are probed by the system loader. However,
+dependencies of a dependency are still searched by their module names only.[3]
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:19]:
```suggestion
dependencies of a dependency are still searched by their module names only [3].
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+So, the risk is minimized only for a dependency itself and not for the libraries
+it depends on, not to mention that the file of a dependency can be replaced in
+the file system by an attacker, which breaks even that protection. Besides that,
+loading of a dependency by specifying full path represents an inconvenience to
+the developers who want to make use of their own variant of a dependency. Not
+only they need to place their variant of a dependency to all of the places from
+which it is going to be found and loaded by every client component that depends
+on it, but also this makes problematic the implementation (if not impossible) of
+some scenarios where being loaded dependency maintains single state shared among
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:28]:
```suggestion
some scenarios where the dependency  being loaded maintains single state shared among
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+all its clients. Such scenarios are hard to implement because copies of the same
+DLL loaded from different paths are considered to be different DLLs and in
+certain cases there is no support for filesystem linking mechanism to point to a
+single file.[4, 5]
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:32]:
```suggestion
single file [4, 5].
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+So, what is the main problem due to which loading by a module name makes Windows
+much more vulnerable to DLL injection than Linux?
+
+Besides difference in the order of accessing paths specified in the environment,
+Windows also prioritizes searching in the directory from which the application
+is loaded and current working directory.[2] Assuming that application is loaded
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:39]:
```suggestion
is loaded and current working directory [2]. Assuming that application is loaded
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+from the directory that requires administrative permission on write, which is
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:40]:
```suggestion
from a directory that requires administrative permission on write, which is
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+usually the case, it is the current working directory that forms the main DLL
+preloading attack scenario.[1]
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:42]:
```suggestion
preloading attack scenario [1].
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+There are approaches to exclude the current working directory from the search
+order. However, for a library to avoid process-wide changes to the search order
+the only viable solution for a run-time loading is to pass
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:46]:
```suggestion
the only viable solution for run-time loading is to pass
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+~LOAD_LIBRARY_SAFE_CURRENT_DIRS~ flag to the ~LoadLibraryEx~ Windows API.[6]
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:47]:
```suggestion
~LOAD_LIBRARY_SAFE_CURRENT_DIRS~ flag to the ~LoadLibraryEx~ Windows API [6].
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+With the removal of the current working directory from loader's consideration,
+the search order on Windows starts having little difference with the search
+order on Linux. The difference includes the order in which directories specified
+in the environment and system directories are considered, and the presence of
+the first step of looking into an application directory on Windows.[2, 7]
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:53]:
```suggestion
the first step of looking into an application directory on Windows [2, 7].
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+
+Since the system environment variables and the environment of other processes
+cannot be changed, the only vulnerable place is an application directory.[8, 9]
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:56]:
```suggestion
cannot be changed, the only vulnerable place is an application directory [8, 9].
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+Because the application can be installed in a directory that does not require
+administrative permissions on write, it still can be started by an account
+having them. Unlike Linux systems, for the process started with administrative
+permissions, the paths specified in the environment and the application
+directory are still considered by the Windows system loader.[2, 7] Therefore, an
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:61]:
```suggestion
directory are still considered by the Windows system loader [2, 7]. Therefore, an
```

File: rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+
+Fortunately, there is a signature verification process that helps validating the
+authenticity of a binary before loading it into process address space and
+starting its execution. This allows making use of the established search order
+while checking that genuine version of the dependency is used. However, not
+loading the binary because of the failed signature verification might not be
+always desired. Especially, during the development phase or for a software
+distributor who does not have the signature with which to sign the binary.
+Therefore, to preserve backward compatibility of such usage models, it is
+essential to have possibility to disable signature verification.
```
Comment (vossmjp) [inline rfcs/proposed/loading-dependencies/loading-dependencies-by-module-name.org:75]:
```suggestion
essential to have the possibility to disable signature verification.
```

**Review by vossmjp — APPROVED** [review]
Looks good to me. Found some minor typos.

**Review by akukanov — APPROVED** [review]

**Pull request merged**